crates/gpui/src/platform/linux/wayland/client.rs

```diff
@@ -2157,3 +2198,332 @@ impl Dispatch<zwp_primary_selection_source_v1::ZwpPrimarySelectionSourceV1, ()>
+                                                    )),
+                                                    modifiers: state.modifiers,
+                                                    touch_phase: TouchPhase::Moved,
+                                                });
+                                            drop(state);
+                                            window.handle_input(input);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                } else if state.active_touch_id == Some(id) && !state.is_scrolling {
+                    // Single touch mouse mode
+                    state.mouse_location = Some(position);
+
+                    if let Some(window) = state.mouse_focused_window.clone() {
+                        let input = PlatformInput::MouseMove(MouseMoveEvent {
+                            position,
+                            pressed_button: state.button_pressed,
+                            modifiers: state.modifiers,
+                        });
+                        drop(state);
+                        window.handle_input(input);
+                    }
+                }
+            }
+            wl_touch::Event::Up { id, .. } => {
+                if state.active_touch_id == Some(id) {
+                    // Primary touch lifted
+                    if state.is_scrolling && state.second_touch_id.is_some() {
+                        // Promote second touch to primary and exit scroll mode
+                        state.active_touch_id = state.second_touch_id.take();
+                        state.touch_location = state.second_touch_location.take();
+                        state.is_scrolling = false;
+                        state.scroll_start_position = None;
+                        state.pinch_start_distance = None;
+                        state.last_pinch_distance = None;
+                        state.pinch_zoom_threshold = 0.0;
+
+                        // Send scroll end event
+                        if let Some(window) = state.mouse_focused_window.clone() {
+                            let position = state.touch_location.unwrap_or_default();
+                            let input = PlatformInput::ScrollWheel(ScrollWheelEvent {
+                                position,
+                                delta: ScrollDelta::Pixels(point(px(0.0), px(0.0))),
+                                modifiers: state.modifiers,
+                                touch_phase: TouchPhase::Ended,
+                            });
+                            state.mouse_location = Some(position);
+                            state.button_pressed = Some(MouseButton::Left);
+                            drop(state);
+                            window.handle_input(input);
+                        }
```

Comment:
Seems like `state.mouse_focused_window` is always `None`, so `MouseUp`/`ScrollWheel`  are never sent on my machine.